### PR TITLE
AddNicheCases MirPass

### PR DESCRIPTION
**Pull request by mikebenfield**
This pass optimizes switches on the discriminant of a niche-optimized enum.

### CODE REVIEW

**Comment by mikebenfield** [comment]
This was part of PR #94075, but it was requested that I split this and the actual niche-filling optimization into two parts.

File: compiler/rustc_const_eval/src/interpret/intrinsics.rs [hunk 1]
```diff
@@ -230,7 +230,7 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
             }
             sym::discriminant_value => {
                 let place = self.deref_operand(&args[0])?;
-                let discr_val = self.read_discriminant(&place.into())?.0;
+                let discr_val = self.read_discriminant(&place.into(), false)?.0;
```
Comment (RalfJung) [inline compiler/rustc_const_eval/src/interpret/intrinsics.rs:233]:
```suggestion
                let discr_val = self.read_discriminant(&place.into(), /*relative*/ false)?.0;
```

File: compiler/rustc_const_eval/src/interpret/validity.rs [hunk 2]
```diff
@@ -704,7 +704,7 @@ impl<'rt, 'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> ValueVisitor<'mir, 'tcx, M>
     ) -> InterpResult<'tcx, VariantIdx> {
         self.with_elem(PathElem::EnumTag, move |this| {
             Ok(try_validation!(
-                this.ecx.read_discriminant(op),
+                this.ecx.read_discriminant(op, false),
```
Comment (RalfJung) [inline compiler/rustc_const_eval/src/interpret/validity.rs:707]:
```suggestion
                this.ecx.read_discriminant(op, /*relative*/ false),
```

File: compiler/rustc_const_eval/src/interpret/visitor.rs [hunk 3]
```diff
@@ -135,7 +135,7 @@ macro_rules! make_value_visitor {
                 &mut self,
                 op: &OpTy<'tcx, M::PointerTag>,
             ) -> InterpResult<'tcx, VariantIdx> {
-                Ok(self.ecx().read_discriminant(op)?.1)
+                Ok(self.ecx().read_discriminant(op, false)?.1)
```
Comment (RalfJung) [inline compiler/rustc_const_eval/src/interpret/visitor.rs:138]:
```suggestion
                Ok(self.ecx().read_discriminant(op, /*relative*/ false)?.1)
```

File: compiler/rustc_const_eval/src/interpret/operand.rs [hunk 4]
```diff
@@ -766,13 +774,18 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
...
                         } else {
                             dataful_variant
+                        };
+
+                        if relative {
+                            (Scalar::from_uint(variant_index_relative, discr_layout.size), variant)
+                        } else {
+                            // Compute the size of the scalar we need to return.
+                            // No need to cast, because the variant index directly serves as
+                            // discriminant and is encoded in the tag.
```
Comment (RalfJung) [inline compiler/rustc_const_eval/src/interpret/operand.rs:784]: Looking at it here I think this should be before the `if` as it applies to both branches.

Also I am honestly not sure what the first line about the 'size of the scalar' is about. That's just `discr_layout.size`. So I think that first line can be removed (it probably refers to an old version of the code), unless someone else can figure out what it is about. :D

**Comment by oli-obk** [comment]
Considering the crater results from the original PR, I think we should run this PR against all the tests that failed

**Comment by RalfJung** [comment]
Wait, this is an optimization, why would it have any impact on crate test failures?

**Comment by JakobDegen** [comment]
This seems like it removes the assignment to `d` without proving that it's unused later on, which is unsound. Not removing the assignment could fix this.

However, I have more general concerns about this going into MIR opts. MIR is generally supposed to be layout independent, and it being pre-monomorphization means that this opt sometimes won't fire for not great reasons. Especially the prospect of changing one of the key types (`Rvalue`) specifically to enable this kind of layout dependent opt is something I do not think is a good idea.

As best I can tell, there's no reason to believe that LLVM won't be able to figure this out on its own - we may indeed emit worse code, but at least this will only affect compile times and not run time of the compiled artifact. If we to avoid emitting the worse code that of course seems like a laudable goal, but imo it belongs in codegen.

**Comment by mikebenfield** [comment]
LLVM certainly does not figure this out currently. And I'm not sure it can in general. I'll have to revisit this when I can sit down and think about it, but when LLVM IR is emitted, information about what are valid values for the niche is lost, so LLVM doesn't know there's only a handful of extra branches it would need to create. 

That said, I take your point about `RValue` and the appropriateness of doing this in MIR. I'll look into whether I can do something similar in codegen.

**Comment by JakobDegen** [comment]
Yeah, I had misunderstood a message in the other PR as implying this was just for compile times. Imo this makes it even more important that this fires consistently though

**Comment by mikebenfield** [comment]
Closing this as it seems it's no longer necessary for performance.

**Pull request closed without merging**